### ISSUE TITLE
Expand the ActiveEffect data model to include statuses, a set of statusId strings for different configured conditions.

### ISSUE THREAD

**Issue by aaclayton** [foundryvtt/foundryvtt#8880]
In V10 the core software supports associating an ActiveEffect with a single configured status via the `flags.core.statusId` flag. This was limited because it was both only a flag (and therefore undocumented) as well as only being able to support effects which apply a single status condition.

V11 adds the `statuses` field (`{Set<string>}`) which contains a set of status IDs which are conferred by the effect. Effects which use the flag are automatically migrated to place the value of that flag in the `statuses` set.

**Comment by dev7355608** [comment]
This change should be considered breaking. Many systems, modules, and macros get/set the `core.statusId` flag.

**Comment by aaclayton** [comment]
> This change should be considered breaking. Many systems, modules, and macros get/set the `core.statusId` flag.

There is an automatic migration applied in the `BaseActiveEffect` class which inserts any existing value in the `core.statusId` flag into the new `statuses` set. Were you encountering any issue where pre-existing `core.statusId` flags were not being migrated in this way? Or was your concern more theoretical in nature?

**Comment by dev7355608** [comment]
I noticed this change when I was testing the Adequate Vision module in V11, which temporarily removes the invisible status in the See Invisibility detection mode: https://github.com/stwlam/adequate-vision/blob/aaeb714fa01a6b23bd5daf1bea629a3eb10c470b/script.js#L411-L422. This, of course, no longer works, and even causes an infinite recursion in this case.

A quick github search reveals a lot of usage of this flag. Most commonly used in expressions like used in Adequate Vision: `effects.find(e => e.getFlag("core", "statusId") == statusId)` or `effects.find(e => e.flags.core?.statusId == statusId)`, which wouldn't work anymore.
- https://github.com/search?q=getFlag%28%22core%22%2C+%22statusId%22%29&type=code
- https://github.com/search?q=setFlag%28%22core%22%2C+%22statusId%22%29&type=code
- https://github.com/search?q=flags.core.statusId&type=code

Some of the systems using it:
- D&D 3.5: https://gitlab.com/search?search=statusId&nav_source=navbar&project_id=32896652&group_id=15490059&search_code=true&repository_ref=master
- PF1e: https://gitlab.com/search?search=statusId&nav_source=navbar&project_id=16774935&group_id=52588190&search_code=true&repository_ref=master
- SWADE: https://gitlab.com/search?search=statusId&nav_source=navbar&project_id=16269883&group_id=9612227&search_code=true&repository_ref=develop
- WFRP4e: https://github.com/moo-man/WFRP4e-FoundryVTT/blob/74cd6b53f2ed40e4d098f1a1a44db651506b2dcc/modules/system/effect-wfrp4e.js#L74-L81
- DSA5: https://github.com/search?q=repo%3APlushtoast%2Fdsa5-foundryVTT+statusId&type=code

Some of the modules using it:
- Monk Active Tiles: https://github.com/ironmonk88/monks-active-tiles/blob/9e005f062807ddfece01b8a83ff01fa01b12a003/actions.js#L2934
- Build-a-Bonus: https://github.com/krbz999/babonus/blob/e58a8ee279a6283b15d1bf438f529e73912d37ba/scripts/filters.mjs#L395-L430
- Token Z: https://github.com/theripper93/token-z/blob/cf98f7282cf96082e754b26d38810df098264a5c/scripts/main.js#L10
- Token Action HUD: https://github.com/Drental/fvtt-tokenactionhud/blob/05b12d24eb72000d4803ddc96afb04204ca90e49/scripts/rollHandlers/swade/swade-base.js#L70
- Always HP: https://github.com/ironmonk88/always-hp/blob/ddeb031b0d1418c3ef2e2b72ba831509a04504bc/alwayshp.js#L106

And the flag was also used in an official example in this article: https://foundryvtt.com/article/system-data-models/ (Enhancing a Data Model).

**Comment by aaclayton** [comment]
I think my main takeaway from your post is that what is missing here is a shim to ensure that *reading* this data via `doc.getFlag("core", "statusId")` maintains backwards compatibility for a couple versions. Is that a fair assessment?

**Comment by dev7355608** [comment]
If that's possible, that would be good. But I'm not sure what the shim should do if the effect has multiple statuses; so the shim's warning message should say that continuing to use the flag may lead to unexcepted results.

Can only `getFlag("core", "statusId")` be shimmed, or would `e.flags.core?.statusId` still work too? This way of accessing the flag is pretty popular (because it's a lot faster): https://github.com/search?q=core%3F.statusId&type=code.

**Comment by aaclayton** [comment]
Yes, it is possible to add backwards compatible shims for this behavior.
![image](https://user-images.githubusercontent.com/2468703/224211877-20492034-afb5-4fe4-a966-f4405483cf2d.png)

**Comment by FloRad** [comment]
Is there a practical example of how this can be used?

**Comment by aaclayton** [comment]
> Is there a practical example of how this can be used?

When creating (or updating) an ActiveEffect you can assign a set of status IDs to that effect.
```js
ActiveEffect.create({name: "Blow to the Head", status: ["disoriented", "staggered"]});
```
When that effect is applied to an `Actor`, the set of all effect statuses is available on the Actor as `actor.statuses`. Statuses defined in `CONFIG.statusEffects` automatically assign these status IDs as well.
